src/protocols/secure_channel/CASESession.cpp

```diff
@@ -1592,74 +1576,72 @@ CHIP_ERROR CASESession::HandleSigma3a(System::PacketBufferHandle && msg)
-        SuccessOrExit(err = decryptedDataTlvReader.GetBytes(work.tbsData3Signature.Bytes(), work.tbsData3Signature.Length()));
+        VerifyOrExit(data.tbsData3Signature.Capacity() >= decryptedDataTlvReader.GetLength(), err = CHIP_ERROR_INVALID_TLV_ELEMENT);
+        data.tbsData3Signature.SetLength(decryptedDataTlvReader.GetLength());
+        SuccessOrExit(err = decryptedDataTlvReader.GetBytes(data.tbsData3Signature.Bytes(), data.tbsData3Signature.Length()));
 
         // Prepare for Step 5/6
         {
-            MutableByteSpan fabricRCAC{ work.rootCertBuf };
+            MutableByteSpan fabricRCAC{ data.rootCertBuf };
             SuccessOrExit(err = mFabricsTable->FetchRootCert(mFabricIndex, fabricRCAC));
-            work.fabricRCAC = fabricRCAC;
+            data.fabricRCAC = fabricRCAC;
             // TODO probably should make SetEffectiveTime static and call closer to VerifyCredentials
             SuccessOrExit(err = SetEffectiveTime());
         }
 
         // Copy remaining needed data into work structure
         {
-            work.validContext = mValidContext;
+            data.validContext = mValidContext;
 
             // initiatorNOC and initiatorICAC are spans into msg_R3_Encrypted
             // which is going away, so to save memory, redirect them to their
             // copies in msg_R3_signed, which is staying around
             TLV::TLVReader signedDataTlvReader;
-            signedDataTlvReader.Init(work.msg_R3_Signed.Get(), work.msg_r3_signed_len);
+            signedDataTlvReader.Init(data.msg_R3_Signed.Get(), data.msg_r3_signed_len);
             SuccessOrExit(err = signedDataTlvReader.Next(TLV::kTLVType_Structure, TLV::AnonymousTag()));
             SuccessOrExit(err = signedDataTlvReader.EnterContainer(containerType));
 
             SuccessOrExit(err = signedDataTlvReader.Next(TLV::kTLVType_ByteString, TLV::ContextTag(kTag_TBSData_SenderNOC)));
-            SuccessOrExit(err = signedDataTlvReader.Get(work.initiatorNOC));
+            SuccessOrExit(err = signedDataTlvReader.Get(data.initiatorNOC));
 
-            if (!work.initiatorICAC.empty())
+            if (!data.initiatorICAC.empty())
             {
                 SuccessOrExit(err = signedDataTlvReader.Next(TLV::kTLVType_ByteString, TLV::ContextTag(kTag_TBSData_SenderICAC)));
-                SuccessOrExit(err = signedDataTlvReader.Get(work.initiatorICAC));
+                SuccessOrExit(err = signedDataTlvReader.Get(data.initiatorICAC));
             }
         }
 
-        SuccessOrExit(err = DeviceLayer::PlatformMgr().ScheduleBackgroundWork(
-                          [](intptr_t arg) { HandleSigma3b(*reinterpret_cast<HandleSigma3Work *>(arg)); },
-                          reinterpret_cast<intptr_t>(&work)));
-        workPtr = nullptr; // scheduling succeeded, so don't delete
-        mExchangeCtxt->WillSendMessage();
-        mState = State::kHandleSigma3Pending;
+        {
```

Comment:
Why the scope here?

Reply:
I might change it to dynamically pick background or foreground, like SendSigma3a. But I can unscope it for now.

Reply:
Unscoped.